### ISSUE TITLE
Better error handling and reduce unwraps across the code base

### ISSUE THREAD

**Issue by twitu** [nautechsystems/nautilus_trader#1848]
# Feature Request

Use the following command in specific crates to find unwrap and expect usage.

Note: unncessary_wraps checks that a function unnecessarily wraps the output in a Result. However it is not foolproof since it doesn't check pub functions.

```
cargo clippy --all-targets --all-features -- -Wclippy::unwrap_used -Wclippy::expect_used -Wclippy::unnecessary_wraps --no-deps
```

Some recommendations -
* Always prefer expect over unwrap
* Failing data that doesn't match spec loudly (panic/assert) is better than returning error, especially if the system cannot recover from it.
* Modeling error is best done when errors can be grouped based on a particular module and crate 

## Indicators

https://github.com/nautechsystems/nautilus_trader/blob/94dc343b8b030eaad4ebd5471a36f7a25b5ab4eb/nautilus_core/indicators/src/average/wma.rs#L56-L73

Only WeightedMovingAverage can return an actual Error value when being created. Because of it all other indicators are forced to return `Result`s as well. Creating a default, logging error or panicing can significantly reduce Result type from spreading across the code base that interacts with Indicators.

https://github.com/nautechsystems/nautilus_trader/blob/94dc343b8b030eaad4ebd5471a36f7a25b5ab4eb/nautilus_core/indicators/src/momentum/amat.rs#L126-L151

Refactor to remove unwraps

## Execution

https://github.com/nautechsystems/nautilus_trader/blob/94dc343b8b030eaad4ebd5471a36f7a25b5ab4eb/nautilus_core/execution/src/engine.rs#L155-L165

Log error and continue or package error into result type

## Common

https://github.com/nautechsystems/nautilus_trader/blob/94dc343b8b030eaad4ebd5471a36f7a25b5ab4eb/nautilus_core/common/src/xrate.rs#L81-L95

Entry is already being created using default. Refactor to remove unwraps.

https://github.com/nautechsystems/nautilus_trader/blob/94dc343b8b030eaad4ebd5471a36f7a25b5ab4eb/nautilus_core/common/src/timer.rs#L135-L153

Validation checks can fail, removing Result from TestTimer code. Similarly for OrderIniitialized and other dependents.

https://github.com/nautechsystems/nautilus_trader/blob/94dc343b8b030eaad4ebd5471a36f7a25b5ab4eb/nautilus_core/common/src/cache/database.rs#L111-L115

Most implementors end up unwrapping cache functions except the one that wraps it into a python error. Logging error might be better if it is not a fatal error.

## Core

https://github.com/nautechsystems/nautilus_trader/blob/94dc343b8b030eaad4ebd5471a36f7a25b5ab4eb/nautilus_core/core/src/datetime.rs#L117

https://github.com/nautechsystems/nautilus_trader/blob/94dc343b8b030eaad4ebd5471a36f7a25b5ab4eb/nautilus_core/core/src/datetime.rs#L141

Use `days` when input is known to be valid at compile time

https://github.com/nautechsystems/nautilus_trader/blob/94dc343b8b030eaad4ebd5471a36f7a25b5ab4eb/nautilus_core/core/src/equality.rs#L24-L29

Equality check only used in test. Move inside test logic or use default false if value can't be serialized.

https://github.com/nautechsystems/nautilus_trader/blob/94dc343b8b030eaad4ebd5471a36f7a25b5ab4eb/nautilus_core/core/src/nanos.rs#L115-L155

Consider using default value or error logs for overflow errors. If panic is not the preferred behaviour.

https://github.com/nautechsystems/nautilus_trader/blob/94dc343b8b030eaad4ebd5471a36f7a25b5ab4eb/nautilus_core/core/src/uuid.rs#L62

Use expect or include Cstr related errors in error domain model.

https://github.com/nautechsystems/nautilus_trader/blob/94dc343b8b030eaad4ebd5471a36f7a25b5ab4eb/nautilus_core/core/src/correctness.rs#L98-L103

Convert correctness checks to expect/panic instead of returning an error. There are a few reasons for this -
* Additional effort in modeling error domain
* Handling things outside the specification is not necessary. Handling it using Result makes the API bloated and all callers need to handle Result
* Debugging is harder because the stack trace only shows the location where the result was unwrapped not where it was created. Which means if `check_string_contains` returns an error and the error gets propagated to higher callers we lose out on information about which set of calls led to the check failing in the first place. The only way to recover that information is through detailed logging.
* Most of the results eventually get unwrapped in model with no other handling making the Result type redundant

## Model

https://github.com/nautechsystems/nautilus_trader/blob/94dc343b8b030eaad4ebd5471a36f7a25b5ab4eb/nautilus_core/model/src/stubs.rs#L48-L52

Validness checks or better domain modeling for Decimals

https://github.com/nautechsystems/nautilus_trader/blob/94dc343b8b030eaad4ebd5471a36f7a25b5ab4eb/nautilus_core/model/src/types/currency.rs#L66-L68

Currency and Venue map locking errors can be modeled better since they are passed to Python. On the other hand the logic could wait to acquire lock with timeout. And if it timesout out, there's clearly a deadlock since these maps aren't under heavy contention.

https://github.com/nautechsystems/nautilus_trader/blob/94dc343b8b030eaad4ebd5471a36f7a25b5ab4eb/nautilus_core/model/src/orderbook/aggregation.rs#L42-L45

The function result is always unwrapped except when it is return to Python.

https://github.com/nautechsystems/nautilus_trader/blob/94dc343b8b030eaad4ebd5471a36f7a25b5ab4eb/nautilus_core/model/src/instruments/synthetic.rs#L89-L94

Change formula returns Result when formula cannot be created. But it is always unwrapped. Better to use expect and panic.

Correctness checks from common cause a lot of unwrapping in model.

**Comment by cjdsellers** [comment]
The first round of changes focused on error handling within the core domain model, emphasizing a "fail fast" approach and making "invalid states unrepresentable" (merged with #1849). 

According to @twitu's notes, it's better to immediately fail when the system would otherwise fall out of spec with no chance of recovery—aligning with the "parse, don't validate" philosophy. This approach necessitates checking of external inputs before passing them to domain objects when issues or bad data are anticipated.

A future improvement will be to identify groups or hierarchies of errors and define more custom `Error` types using `thiserror`.

**Comment by twitu** [comment]
With two rounds of refactoring, unncessary Result types, unwraps and expects have been been reduced significantly. From the warnings identified in the previous post most have been refactored and the below are remaining.

The rest can be identified and resolved on an individual basis as and when encountered.

## Common

https://github.com/nautechsystems/nautilus_trader/blob/94dc343b8b030eaad4ebd5471a36f7a25b5ab4eb/nautilus_core/common/src/cache/database.rs#L111-L115

Most implementors end up unwrapping cache functions except the one that wraps it into a python error. Logging error might be better if it is not a fatal error.

## Model

https://github.com/nautechsystems/nautilus_trader/blob/94dc343b8b030eaad4ebd5471a36f7a25b5ab4eb/nautilus_core/model/src/stubs.rs#L48-L52

Validness checks or better domain modeling for Decimals

https://github.com/nautechsystems/nautilus_trader/blob/94dc343b8b030eaad4ebd5471a36f7a25b5ab4eb/nautilus_core/model/src/types/currency.rs#L66-L68

Currency and Venue map locking errors can be modeled better since they are passed to Python. On the other hand the logic could wait to acquire lock with timeout. And if it timesout out, there's clearly a deadlock since these maps aren't under heavy contention.

https://github.com/nautechsystems/nautilus_trader/blob/94dc343b8b030eaad4ebd5471a36f7a25b5ab4eb/nautilus_core/model/src/orderbook/aggregation.rs#L42-L45

The function result is always unwrapped except when it is return to Python.

https://github.com/nautechsystems/nautilus_trader/blob/94dc343b8b030eaad4ebd5471a36f7a25b5ab4eb/nautilus_core/model/src/instruments/synthetic.rs#L89-L94

Change formula returns Result when formula cannot be created. But it is always unwrapped. Better to use expect and panic.
